File: 1277. Count Square Submatrices with All Ones - Medium/solution.cpp
static auto _ = [] () { ios_base::sync_with_stdio(false); cin.tie(nullptr); return 0; } ();
class Solution {
public:
    int countSquares(vector < vector < int > > & matrix) {
        const int n = matrix.size();
        const int m = matrix[0].size();
        
        int dp[2][m];
        int result = 0;
        
        int c = 1;
        for (int i = 0; i < n; ++i) {
            for (int j = 0; j < m; ++j) {
                dp[c][j] = matrix[i][j];
                if (i != 0 && j != 0 && dp[c][j]) {
                    dp[c][j] += min(min(dp[c^1][j], dp[c^1][j-1]), dp[c][j-1]);
                }
                
                result += dp[c][j];
            }
            
            c ^= 1;
        }
        
        return result;
    }
};

File: 1277. Count Square Submatrices with All Ones - Medium/solution.py
class Solution:
    def countSquares(self, matrix: List[List[int]]) -> int:
        if matrix is None or len(matrix) == 0:
            return 0

        result = 0
        for r in range(len(matrix)):
            for c in range(len(matrix[0])):
                if matrix[r][c] == 1:
                    if r == 0 or c == 0:
                        result += 1
                    else:
                        cell_val = min(matrix[r-1][c-1], matrix[r][c-1], matrix[r-1][c]) + matrix[r][c]
                        result += cell_val
                        matrix[r][c] = cell_val
        return result